File: server-nestjs-single/src/scopes/CommentScope.ts
import type { RhinoScope } from '@rhino-dev/rhino-nestjs';

/**
 * CommentScope — single-tenant ownership isolation (inherited via Task → Project).
 *
 * A comment is owned through its task's project. Constrain to comments whose
 * task belongs to a project owned by the current user.
 */
export class CommentScope implements RhinoScope {
  apply(where: Record<string, any>, ctx: { user?: any }) {
    if (ctx.user?.id) {
      return {
        ...where,
        task: { ...(where.task ?? {}), project: { userId: ctx.user.id } },
      };
    }
    return where;
  }
}

File: server-rails-single/app/models/scopes/comment_scope.rb
# frozen_string_literal: true

module Scopes
  # CommentScope — single-tenant ownership isolation (inherited via Task → Project).
  #
  # A comment is owned through its task's project. We constrain to comments whose
  # task belongs to a project owned by the current user.
  class CommentScope < Rhino::ResourceScope
    def apply(relation)
      return relation unless user

      owned_projects = Project.unscoped.where(user_id: user.id).select(:id)
      owned_tasks = Task.unscoped.where(project_id: owned_projects).select(:id)
      relation.where(task_id: owned_tasks)
    end
  end
end
